ocode_python/core/stream_processor.py

```
@@ -0,0 +1,553 @@
+            )
+
+    async def _execute_read_operation(self, operation: Operation) -> ToolResult:
+        """Execute a read operation with caching."""
+        async with self._read_lock:
+            cache_key = f"{operation.tool_name}:{hash(str(operation.arguments))}"
+
+            if cache_key in self.read_cache:
+                return self.read_cache[cache_key]
+
+            # Execute the read operation
+            from ..tools.base import ToolRegistry
```

Pylint (3.3.7)

[convention] Import outside toplevel (tools.base.ToolRegistry)

(C0415)